### PR TITLE
fix(inbound): handle media content types

### LINKED ISSUE #11: Fix inbound Telegram media uploads failing due to disallowed content_type (octet-stream)
## User request
Telegram connector works for text messages, but when sending an image the connector logs repeated errors like:

```text
connector: handle update error: finalize upload: invalid_argument: content_type is not allowed
connector: handle update error: finalize upload: internal: EOF
```

Goal: inbound Telegram images (and other media) should upload successfully to the Files service and not spam errors / get stuck.

### Root cause
The connector’s upload path currently prefers the **Telegram file download HTTP header** `Content-Type` when present. Telegram frequently returns `Content-Type: application/octet-stream` for file downloads, even when the update includes a correct MIME type (or when we already know the type, e.g. photos). Our Files service rejects `application/octet-stream` (`INVALID_ARGUMENT content_type is not allowed`), and depending on timing the stream may be cut early which can surface as `EOF`.

### Required changes
1. **Determine upload `content_type` using an “allowed-aware” selection step** before starting the Files upload stream.
   - Normalize any candidate via `mime.ParseMediaType()`, drop params, lowercase.
   - Treat `application/octet-stream` as “unknown” and do not select it unless no other valid option exists.
   - Candidate precedence:
     1) Telegram update mime type (for photos: default `image/jpeg`)
     2) Sniffed type from bytes (e.g. `http.DetectContentType`)
     3) Extension-derived type from Telegram `file_path` / filename
     4) Telegram download header `Content-Type`
   - Choose the first candidate that is non-empty, not generic, and passes the same allowlist rules as the Files service (image/*, text/*, video/*, audio/*, plus the specific `application/*` types).

2. **Photo-specific requirement**
   - For `message.photo`, final chosen `content_type` must be `image/*` and must not be `application/octet-stream`.
   - If all detection fails, force `image/jpeg` as a last resort.

3. **Alias mapping (minimal required set)**
   - `application/x-zip-compressed` → `application/zip`
   - `application/x-gzip` → `application/gzip`

4. **Pre-flight validation**
   - If the computed `content_type` is still not allowed, do **not** call Files `UploadFile()`.
   - Ensure inbound polling continues (don’t block offset advancement). Prefer a consistent “skip attachment” behavior + an audit/status log entry (include file_id + candidate content-types observed).

5. **Error handling expectations**
   - `INVALID_ARGUMENT content_type is not allowed`: non-retriable unless retrying with a *different* computed `content_type`.
   - Transport-like errors (`EOF`, `UNAVAILABLE`, `DEADLINE_EXCEEDED`, generic `INTERNAL`): retry upload up to 3 attempts with exponential backoff (new stream each attempt).

### Tests
- Unit tests for content-type selection:
  - Header octet-stream but telegram mime present → choose telegram mime.
  - Header octet-stream, telegram mime empty, sniff works → choose sniffed `image/*`.
  - Alias mapping for zip/gzip.
  - Disallowed candidates → upload not attempted and attachment skipped.
- E2E inbound test:
  - Mock Telegram photo download response to return `Content-Type: application/octet-stream` and confirm upload still succeeds with `image/jpeg` (or sniffed `image/*` if used).

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- add allowed-aware content-type selection for inbound media uploads with retries and skip/audit logging
- add unit coverage for content-type selection and skip behavior
- update inbound e2e media test for octet-stream headers

## Testing
- go vet ./...
- go test ./...
- go test -tags e2e ./test/e2e

Closes #11

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Summary
- add allowed-aware content-type selection for inbound uploads with retry + skip auditing
- add content-type selection unit tests and update inbound e2e media coverage for octet-stream headers

## Testing
- go vet ./...
- go test ./...
- go test -tags e2e ./test/e2e

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- go vet ./... (lint: no issues)
- go test ./... (passed: all, failed: 0, skipped: 0)
- go test -tags e2e ./test/e2e (passed: all, failed: 0, skipped: 0)

File: internal/connector/inbound.go [hunk 1]
```diff
@@ -307,39 +309,39 @@ func (w *installationWorker) buildInboundMessage(ctx context.Context, message *t
 	caption := strings.TrimSpace(message.Caption)
 	if len(message.Photo) > 0 {
 		photo := largestPhoto(message.Photo)
-		fileID, err := w.uploadTelegramFile(ctx, photo.FileID, "photo", "image/jpeg")
+		fileID, err := w.uploadTelegramFile(ctx, photo.FileID, "photo", "image/jpeg", true)
 		if err != nil {
 			return "", nil, err
 		}
-		return caption, []string{fileID}, nil
+		return caption, fileIDsOrNil(fileID), nil
```
Comment (noa-lucent) [inline internal/connector/inbound.go]: [major] When uploadTelegramFile skips an attachment it returns ""/nil, so fileIDsOrNil(fileID) becomes nil. If the caption is also empty (common for photo-only messages), buildInboundMessage returns ("", nil, nil) and Gateway.SendMessage will fail (see internal/gateway/client.go: body or files required). That will block offset advancement and reintroduce the "stuck/spam errors" behavior.\n\nPlease ensure we never call SendMessage when both body and fileIDs are empty (e.g. return a placeholder body like "[attachment skipped]" from buildInboundMessage, or short-circuit handleUpdate). Add a unit test for the empty-caption + skipped-attachment case.

File: internal/connector/inbound.go [hunk 2]
```diff
@@ -351,34 +353,126 @@ func (w *installationWorker) buildInboundMessage(ctx context.Context, message *t
...
+			continue
+		}
 		return "", err
 	}
-	return uploaded.GetId(), nil
+	if contentTypeErr != nil {
+		w.logAttachmentSkipped(ctx, fileID, candidates)
+		return "", nil
+	}
+	return "", fmt.Errorf("upload file: no content type selected")
```
Comment (noa-lucent) [inline internal/connector/inbound.go]: [minor] This "no content type selected" return looks unreachable: allowedContentTypes is non-empty, and the loop either returns success, returns an error, or ends with contentTypeErr!=nil and skips. Consider removing this branch (or replacing with a clear assert/panic) to keep invariants crisp.

File: internal/connector/inbound.go [hunk 3]
```diff
@@ -351,34 +353,126 @@ func (w *installationWorker) buildInboundMessage(ctx context.Context, message *t
...
+		if !isUploadRetriableError(err) || attempts >= inboundUploadRetryMax {
+			return nil, err
+		}
+		if !sleepContext(ctx, retryDelay(attempts)) {
+			return nil, ctx.Err()
+		}
+	}
+}
+
+func isUploadContentTypeNotAllowed(err error) bool {
```
Comment (noa-lucent) [inline internal/connector/inbound.go:448]: [minor] isUploadContentTypeNotAllowed() currently keys off strings.Contains(err.Error(), "content_type is not allowed") and returns true even when the error isn’t a *connect.Error. To avoid misclassifying unrelated errors (and accidentally skipping attachments), prefer: errors.As(err, &connectErr) && connectErr.Code()==connect.CodeInvalidArgument && strings.Contains(strings.ToLower(connectErr.Message()), "content_type is not allowed"). Only fall back to err.Error() matching if we truly see non-connect errors here.

File: internal/connector/content_type.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+import (
+	"fmt"
+	"mime"
+	"path"
+	"strings"
+)
+
+const photoFallbackContentType = "image/jpeg"
+
+var allowedContentTypePrefixes = []string{
```
Comment (noa-lucent) [inline internal/connector/content_type.go:13]: [minor] Please sanity-check that allowedContentTypePrefixes matches the Files service allowlist (so our preflight skip behavior is accurate). Also note we’re using HasPrefix for exact types like application/json/application/xml; if we want strict matching, consider splitting into prefix rules (image/, text/, …) + an exact set for application/* types.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Blocking: skip-path can produce (body="", fileIDs=nil) for media messages without captions, which fails SendMessage (body or files required) and prevents offset advancement (stuck/spam again). Please ensure skipped attachments never lead to an empty message being sent (guard in handleUpdate or return a placeholder body), and add a unit test for that case.\n\nNon-blocking: tighten isUploadContentTypeNotAllowed matching, remove unreachable fallback branch, and double-check preflight allowlist fidelity.

**Comment by casey-brooks** [comment]
## Summary
- skip sending inbound messages when both body and file IDs are empty after attachment skip
- tighten content_type error detection and align allowlist note
- add unit coverage for empty-body skipped-attachment case

## Testing
- go vet ./...
- go test ./...
- go test -tags e2e ./test/e2e

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- go vet ./... (lint: no issues)
- go test ./... (passed: all, failed: 0, skipped: 0)
- go test -tags e2e ./test/e2e (passed: all, failed: 0, skipped: 0)

**Review by noa-lucent — DISMISSED** [review]
Re-review: blocker addressed. handleUpdate now skips sending when both body and fileIDs are empty, preventing offset advancement from getting stuck; added unit test covers attachment-skip + empty-caption case. The content_type error detection and unreachable branch cleanup look good.

**Comment by casey-brooks** [comment]
## Summary
- emit user-visible attachment-skip markers, appended to captions or used as standalone body
- include filename/size/content type in markers and return markers from skip handling
- extend unit coverage for skip markers in upload + inbound send cases

## Testing
- go vet ./...
- go test ./...
- go test -tags e2e ./test/e2e

File: internal/connector/inbound.go [hunk 5]
```diff
@@ -351,34 +357,160 @@ func (w *installationWorker) buildInboundMessage(ctx context.Context, message *t
...
+	w.appendAudit(ctx, auditEvent{
+		name:           auditEventAttachmentSkipped,
+		message:        message,
+		level:          appsv1.InstallationAuditLogLevel_INSTALLATION_AUDIT_LOG_LEVEL_WARNING,
+		idempotencyKey: auditKeyWithHash(auditEventAttachmentSkipped, w.installation.ID, fmt.Sprintf("%s:%s", fileID, summary)),
+	})
+	return formatAttachmentSkippedMarker(filename, sizeBytes, candidates, summary)
+}
+
+func formatAttachmentSkippedMarker(filename string, sizeBytes int64, candidates []contentTypeCandidate, summary string) string {
```
Comment (noa-lucent) [inline internal/connector/inbound.go:495]: [minor] The new user-visible skip marker is great, but it embeds untrusted strings (filename/content-type). To avoid reintroducing a "stuck" scenario if Threads rejects an overly-long/odd body, please sanitize + bound these fields (e.g., replace newlines with spaces and truncate to a reasonable max length) before formatting the marker.

**Review by noa-lucent — APPROVED** [review]
Re-review: the new user-visible "[attachment skipped …]" marker behavior looks correct (marker sent when attachment is skipped and appended to caption when present), and unit tests were updated accordingly. Left one minor follow-up: sanitize/truncate marker fields to keep the body bounded and avoid any chance of SendMessage failing on extreme filenames/content-types.

File: internal/connector/inbound.go (hunk 5 above)
Comment (noa-lucent) [reply internal/connector/inbound.go:495]: Non-blocking suggestion (ok to address in a follow-up). Resolving so it doesn’t block approval.
